#576 Unique Id Mapper should not return an option

anilgursel opened the pull request:
Thanks for your pull request.  Please review the following guidelines.

- [x] Title includes issue id.
- [x] Description of the change added.
- [x] Commits are squashed.
- [ ] Tests added.
- [x] Documentation added/updated.
- [x] Also please review [CONTRIBUTING.md](https://github.com/paypal/squbs/blob/master/CONTRIBUTING.md).

akara commented on squbs-ext/src/main/scala/org/squbs/streams/RetryBidi.scala:
```diff
@@ -71,7 +71,7 @@ object RetryBidi {
     * @return a [[BidiFlow]] with Retry functionality
     */
-  def apply[In, Out, Context](maxRetries: Int, uniqueIdMapper: Context => Option[Any] = (_: Any) => None,
+  def apply[In, Out, Context](maxRetries: Int, uniqueIdMapper: Context => Any = (c: Context) => c,
```
I think it is more clear and compact if you say `uniqueIdMapper: Context => Any = identity`. Please try it out.

akara commented on squbs-ext/src/main/scala/org/squbs/streams/TimeoutBidi.scala:
```diff
@@ -300,11 +300,9 @@ final class TimeoutBidiUnordered[In, Out, Context](timeout: FiniteDuration,
-      }
+    uniqueIdMapper(context) match {
+      case uniqueIdProvider: UniqueId.Provider ⇒ uniqueIdProvider.uniqueId
+      case uniqueId ⇒ uniqueId
```
One reason for getting rid of `Context => Option[Any]` is to optimize the decision to be just once. I don't think we have achieved this just yet. If you would model it as `Option[Context => Any]` then you could make this decision upfront. Does not mean the API needs to show `Option` or `Optional`. That we can do behind the scenes instead of a function that gets called all the time. Just for consideration...

anilgursel replied:
The only thing we could save is the `match`, we still need to call the function.  Because the `match` won't change based on the request.

However, do we want to optimize not calling `uniqueIdMapper` when one not provided?  I would assume if we set it to `identity`, compiler would do the optimization.

akara replied:
Yes, I also assume `identity` will reach that same result. Compiler should optimize it away. So if we can make sure the match does not happen for every message, I'm good.

akara commented on squbs-ext/src/main/scala/org/squbs/streams/UniqueId.scala:
```diff
@@ -37,11 +37,8 @@ object UniqueId {
-
-    (context: Context) => toScala(uniqueIdMapper(context))
+  private[squbs] def javaUniqueIdMapperAsScala[Context](uniqueIdMapper: JFunction[Context, Any]):
+  Context => Any = {
```
`... Context => Any = uniqueIdMapper.apply`

May need an `_` after the apply just in case the compiler does not recognize it.

anilgursel replied:
I had actually eliminated all uses of this function, but apparently, forgot to delete it.

akara commented on squbs-ext/src/main/scala/org/squbs/streams/circuitbreaker/japi/CircuitBreakerSettings.scala:
```diff
@@ -60,7 +60,7 @@ case class CircuitBreakerSettings[In, Out, Context] private[japi] (
       (out: Out) => cleanUp.accept(out),
       failureDecider.asScala.map(f => (out: Try[Out]) => f(out).asInstanceOf[Boolean]),
-      UniqueId.javaUniqueIdMapperAsScala(uniqueIdMapper))
+      uniqueIdMapper.asScala)
```
Ditto. You don't need a function converter to convert a Java 8 `Function` to Scala. Just the `Function.apply _` is by itself a Scala function.

anilgursel commented:
@akara Please check if the last commit addresses the concerns.  Please note, I am not very happy how the API currently looks and also, there is some inconsistency with `TimeoutBidiUnordered` (which takes an `Option`).   However, I will address the concerns with another PR as follows:

* Rename stages as `Retry`, `CircuitBreaker`, `Timeout`.
* Each takes a `Settings`, defined inside the corresponding companion object.
* Retry, CircuitBreaker and Timeout will have APIs that take settings, but also their simple form, e.g., Retry(5).

akara commented on docs/flow-timeout.md:
````diff
@@ -126,7 +126,7 @@ The following API is used to pass a uniqueId mapper:
 ```scala
 TimeoutBidiFlowUnordered[In, Out, Context](timeout: FiniteDuration,
-                                           uniqueIdMapper: Context => Option[Any])
+                                           uniqueIdMapper: OPtion[Context => Any])
````
Eh? Typo in `OPtion`?

akara commented on squbs-ext/src/main/scala/org/squbs/streams/RetryBidi.scala:
```diff
@@ -204,13 +163,12 @@ final class RetryBidi[In, Out, Context] private[streams](maxRetries: Int, unique
-        case uniqueIdProvider: UniqueId.Provider ⇒ uniqueIdProvider.uniqueId
-        case `context` ⇒ `context`
-      }
+  val uniqueId: Context => Any = uniqueIdMapper.getOrElse{
```
Yes, I do like this much more. Thanks!

akara commented on squbs-ext/src/main/scala/org/squbs/streams/TimeoutBidi.scala:
```diff
@@ -299,13 +298,12 @@ final class TimeoutBidiUnordered[In, Out, Context](timeout: FiniteDuration,
-        case uniqueIdProvider: UniqueId.Provider ⇒ uniqueIdProvider.uniqueId
-        case context ⇒ context
-      }
+  val uniqueId: Context => Any = uniqueIdMapper.getOrElse{
```
👍

akara commented on squbs-ext/src/main/scala/org/squbs/streams/RetryBidi.scala:
```diff
@@ -423,14 +381,14 @@ final class RetryBidi[In, Out, Context] private[streams](maxRetries: Int, unique
 case class RetrySettings[In, Out, Context] private[streams](
    maxRetries: Int,
-   uniqueIdMapper: Context => Option[Any] = (_: Any) => None,
+   uniqueIdMapper: Option[Context => Any] = None,
```
We wanted to rename `RetrySettings` to `Retry.Settings`, right? Is that supposed to be in a different PR?

anilgursel replied:
Yes, that will be a separate PR.
